geode-core/src/main/java/org/apache/geode/internal/net/BufferPool.java

```diff
@@ -43,8 +44,8 @@ public BufferPool(DMStats stats) {
   /**
    * A list of soft references to byte buffers.
    */
-  private final ConcurrentLinkedQueue<BBSoftReference> bufferQueue =
-      new ConcurrentLinkedQueue<>();
+  private final ConcurrentLinkedDeque<ByteBufferSoftReference> bufferQueue =
```

Comment:
The reason for switching to the deque was so that we can add it back to the front. If the connection was added back to the end, then we would end up looping over the last buffer, which wouldn't be big enough. This way, we do only one pass over the list of buffers and when we reach the end of the iterator, we can know that we have checked them all.